docs/proposals/14-wireguard-based-cluster-vpn.md

```diff
@@ -0,0 +1,178 @@
+
+# Proposal
+
+The general idea of the proposal is to keep the existing cluster vpn solution more or less as is, but change the underlying network used for the `vpn seed => vpn shoot` connection. The underlying network should be established in the reversed direction, i.e. the shoot cluster should initiate the network connection, but it nevertheless should work in both directions.
+
+We achieve this by tunneling the open vpn connection through a WireGuard tunnel, which is established from the shoot to the seed (note that WireGuard uses udp as protocol). Independent of that we can also use udp for the open vpn connection, but we can also stay with tcp as it was before. While this might look like a big change, it only introduces minor changes to the existing solution, but let's look at the details. In essence, the OpenVPN connection does not require a public endpoint in the shoot cluster but it used the internal endpoint provided by the Wireguard tunnel.
+
+The WireGuard network needs a separate network range/cidr. It has to be unique for the seed and all its shoot clusters. An example for an assumed workload of around 250 shoot clusters with enough buffer would be 192.168.0.0/22 (1024 IP addresses), i.e. 192.168.0.0-192.168.3.255. The IP addresses from this range need to be managed, but the IP address management (IPAM) using the gardener Kubernetes objects like seed and shootstate as backing store is fairly straight forward. This is especially true as we do not expect large network ranges and only infrequent IP allocations. Hence, the IP address allocation can be quite simple, i.e. scan the range for a free IP address.
+
+There is another restriction: in case shoot clusters are configured to be seed clusters this network range must not overlap with the "parent" seed cluster. If the parent seed cluster uses `192.168.0.0/22` the child seed cluster can for example use `192.168.4.0/22`. Grandchildren can however use grandparent IP address ranges.
+
+This slightly adds to the restrictions described in the current solution outline. In that the arbitrary chosen `192.168.123.0/24` range is resticted. For the purpose of this implementation we propose to extend that restriction to `192.168.128.0/17` range. Most of it would be reserved for "future use however".
+
+WireGuard can utilize the Linux kernel so that after initialization/configuration no user space processes are required. We propose to recommend the wireguard kernel module as the default solution for all seeds. For shoot clusters, the WireGuard kernel based approach is also recommended, but the user space solution should also work as we expect less traffic on the shoot side. We expect the userspace implementation to work on all operating systems supported by Gardener in case no kernel module is available.
```

Comment:
> We propose to recommend the wireguard kernel module as the default solution for all seeds.

I don't like this. We don't have any requirements to the seeds other than being a "Kubernetes conformance test"-passing cluster. Now you want to introduce additional requirements/recommendations about the host OS configuration (which sometimes is not even controllable).
So, I guess you must rather assume that this kernel module is not available (neither in seeds nor in shoots).

Reply:
Sorry, I did not consider even my own requirements here (e.g. vSphere supervisor cluster). Can we agree on the following wording:

For seed clusters hosting many control planes (>50) we recommend to use an operating system which provides a WireGuard kernel module.

Reply:
Maybe. I would be interested in the performance difference and how crucial it would be to follow this recommendation for productive Gardener landscapes. Generally, if we could find a solution without such the necessity of considering such additional recommendations, I would be more happy as it would reduce complexity and cognitive overhead for operators (it's already complex to operate and we should prevent making it more complex if possible).